File: sdk/Networking/DataConnectionUtilities.m
//
//  DataConnectionUtilities.m
//  mage-ios-sdk
//
//  Created by Daniel Barela on 2/7/20.
//  Copyright © 2020 National Geospatial-Intelligence Agency. All rights reserved.
//

#import <SystemConfiguration/CaptiveNetwork.h>
#import <SystemConfiguration/SystemConfiguration.h>
#import <CoreTelephony/CTTelephonyNetworkInfo.h>
#import <CoreTelephony/CTCarrier.h>
#import "DataConnectionUtilities.h"

@implementation DataConnectionUtilities

+ (ConnectionType)connectionType {
    SCNetworkReachabilityRef reachability = SCNetworkReachabilityCreateWithName(NULL, "8.8.8.8");
    SCNetworkReachabilityFlags flags;
    BOOL success = SCNetworkReachabilityGetFlags(reachability, &flags);
    CFRelease(reachability);
    if (!success) {
        return ConnectionTypeUnknown;
    }
    BOOL isReachable = ((flags & kSCNetworkReachabilityFlagsReachable) != 0);
    BOOL needsConnection = ((flags & kSCNetworkReachabilityFlagsConnectionRequired) != 0);
    BOOL isNetworkReachable = (isReachable && !needsConnection);

    if (!isNetworkReachable) {
        return ConnectionTypeNone;
    } else if ((flags & kSCNetworkReachabilityFlagsIsWWAN) != 0) {
        //connection type
        CTTelephonyNetworkInfo *netinfo = [[CTTelephonyNetworkInfo alloc] init];
        NSDictionary *carrier = [netinfo serviceSubscriberCellularProviders];
        NSDictionary *radio = [netinfo serviceCurrentRadioAccessTechnology];

        NSLog(@"Carrier %@", carrier);
        NSLog(@"Radio %@", radio);

        return ConnectionTypeCell;
    } else {
        return ConnectionTypeWiFi;
    }
}

+ (NSString *) getCurrentWifiSsid {
    NSString *wifiName;
    enum ConnectionType type = [DataConnectionUtilities connectionType];
    if (type == ConnectionTypeWiFi) {
        CFArrayRef interfaces = CNCopySupportedInterfaces();
        if (interfaces) {
            CFIndex count = CFArrayGetCount(interfaces);
            for (int i = 0; i < count; i++) {
                CFStringRef interface = (CFStringRef)CFArrayGetValueAtIndex(interfaces, i);
                NSLog(@"Interface %@", interface);
                NSDictionary *dictionary = (__bridge NSDictionary*)CNCopyCurrentNetworkInfo(interface);
                NSLog(@"Dictionary %@", dictionary);
                // if dictionary is nil then there is no wifi
                if (dictionary) {
                    wifiName = [NSString stringWithFormat:@"%@",[dictionary objectForKey:(__bridge NSString *)kCNNetworkInfoKeySSID]];
                }
            }
            CFRelease(interfaces);
        }
    }
    return wifiName;
}

+ (BOOL) currentWiFiAllowed {
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    NSInteger wifiNetworkRestrictionType = [[defaults objectForKey:@"wifiNetworkRestrictionType"] longValue];
    if (wifiNetworkRestrictionType == WIFIRestrictionTypeNoRestrictions) {
        return true;
    }
    
    NSString *currentSSID = [DataConnectionUtilities getCurrentWifiSsid];
    if (wifiNetworkRestrictionType == WIFIRestrictionTypeOnlyTheseWifiNetworks) {
        NSArray *whitelist = [defaults objectForKey:@"wifiWhitelist"];
        if (currentSSID && [whitelist containsObject:currentSSID]) {
            return true;
        }
    } else if (wifiNetworkRestrictionType == WIFIRestrictionTypeNotTheseWifiNetworks) {
        NSArray *blacklist = [defaults objectForKey:@"wifiBlacklist"];
        if (!currentSSID || ![blacklist containsObject:currentSSID]) {
            return true;
        }
    }
    
    return false;
}

+ (BOOL) shouldPerformNetworkOperation: (NSString *) preferencesKey {
    // 0 = all 1 = wifionly 2 = none
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    NSInteger observationPushNetworkOption = [[defaults objectForKey:preferencesKey] longValue];
    if (!observationPushNetworkOption || observationPushNetworkOption == NetworkAllowTypeAll) {
        return true;
    } else if (observationPushNetworkOption == NetworkAllowTypeWiFiOnly) {
        if ([DataConnectionUtilities connectionType] == ConnectionTypeWiFi) {
            return [DataConnectionUtilities currentWiFiAllowed];
        } else {
            return false;
        }
    } else if (observationPushNetworkOption == NetworkAllowTypeNone) {
        return false;
    }
    return false;
}

+ (BOOL) shouldPushObservations {
    return [DataConnectionUtilities shouldPerformNetworkOperation:@"observationPushNetworkOption"];
}

+ (BOOL) shouldFetchObservations {
    return [DataConnectionUtilities shouldPerformNetworkOperation:@"observationFetchNetworkOption"];
}

+ (BOOL) shouldFetchLocations {
    return [DataConnectionUtilities shouldPerformNetworkOperation:@"locationFetchNetworkOption"];
}

+ (BOOL) shouldPushLocations {
    return [DataConnectionUtilities shouldPerformNetworkOperation:@"locationPushNetworkOption"];
}

+ (BOOL) shouldPushAttachments {
    return [DataConnectionUtilities shouldPerformNetworkOperation:@"attachmentPushNetworkOption"];
}

+ (BOOL) shouldFetchAttachments {
    return [DataConnectionUtilities shouldPerformNetworkOperation:@"attachmentFetchNetworkOption"];
}

+ (BOOL) shouldFetchAvatars {
    return [DataConnectionUtilities shouldPerformNetworkOperation:@"attachmentFetchNetworkOption"];
}

@end

File: sdk/Networking/DataConnectionUtilities.swift
//
//  DataConnectionUtilities.m
//  mage-ios-sdk
//
//  Created by Daniel Barela on 2/7/20.
//  Copyright © 2020 National Geospatial-Intelligence Agency. All rights reserved.
//

import Foundation
import SystemConfiguration
import SystemConfiguration.CaptiveNetwork

@objc public enum ConnectionType : Int {
    case unknown
    case none
    case cell
    case wiFi
}

@objc public enum WIFIRestrictionType : Int {
    case NoRestrictions
    case OnlyTheseWifiNetworks
    case NotTheseWifiNetworks
}

@objc public enum NetworkAllowType : Int {
    case all
    case wiFiOnly
    case none
}

@objc class DataConnectionUtilities : NSObject {
    
    
    @objc public static func connectionType() -> ConnectionType {
        
        var zeroAddress = sockaddr_in()
        zeroAddress.sin_len = UInt8(MemoryLayout<sockaddr_in>.size)
        zeroAddress.sin_family = sa_family_t(AF_INET)
        
        guard let defaultRouteReachability = withUnsafePointer(to: &zeroAddress, {
            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                SCNetworkReachabilityCreateWithAddress(nil, $0)
            }
        }) else {
            return .none
        }
        
        var flags: SCNetworkReachabilityFlags = []
        if !SCNetworkReachabilityGetFlags(defaultRouteReachability, &flags) {
            return .none
        }
        
        let isReachable = flags.contains(.reachable)
        
        if (!isReachable) {
            return .none
        } else if (flags.contains(.isWWAN)) {
            return .cell
        } else {
            return .wiFi
        }
    }
    
    @objc public static func getCurrentWifiSsid() -> String? {
        if DataConnectionUtilities.connectionType() != .wiFi {
            return nil
        }
        
        guard let interfaceNames = CNCopySupportedInterfaces() as? [String] else {
            return nil
        }
        let ssids: [String] = interfaceNames.compactMap { name in
            guard let info = CNCopyCurrentNetworkInfo(name as CFString) as? [String:AnyObject] else {
                return nil
            }
            guard let ssid = info[kCNNetworkInfoKeySSID as String] as? String else {
                return nil
            }
            return ssid
        }
        
        if !ssids.isEmpty {
            return ssids[0]
        }
        return nil
    }
    
    static func currentWiFiAllowed() -> Bool {
        let wifiNetworkRestrictionType = UserDefaults.standard.wifiNetworkRestrictionType
        
        if wifiNetworkRestrictionType == .NoRestrictions {
            return true
        }
        
        let currentSSID = DataConnectionUtilities.getCurrentWifiSsid()
        if (wifiNetworkRestrictionType == .OnlyTheseWifiNetworks) {
            if let currentSSID = currentSSID, UserDefaults.standard.wifiWhitelist.contains(currentSSID) {
                return true
            }
        } else if (wifiNetworkRestrictionType == .NotTheseWifiNetworks) {
            if let currentSSID = currentSSID {
                return !UserDefaults.standard.wifiBlacklist.contains(currentSSID)
            } else {
                return true
            }
        }
        return false
    }
    
    static func shouldPerformNetworkOperation(_ preferencesKey: String) -> Bool {
        let networkOption = UserDefaults.standard.integer(forKey: preferencesKey)
        guard let networkAllowType = NetworkAllowType(rawValue: networkOption) else {
            return false
        }
        
        if networkAllowType == .all {
            return true
        } else if networkAllowType == .wiFiOnly {
            if DataConnectionUtilities.connectionType() == .wiFi {
                return DataConnectionUtilities.currentWiFiAllowed()
            } else {
                return false
            }
        } else if networkAllowType == .none {
            return false
        }
        return false
    }
    
    @objc public static func shouldPushObservations() -> Bool {
        return DataConnectionUtilities.shouldPerformNetworkOperation("observationPushNetworkOption");
    }
    
    @objc public static func shouldFetchObservations() -> Bool {
        return DataConnectionUtilities.shouldPerformNetworkOperation("observationFetchNetworkOption");
    }
    
    @objc public static func shouldPushLocations() -> Bool {
        return DataConnectionUtilities.shouldPerformNetworkOperation("locationPushNetworkOption");
    }
    
    @objc public static func shouldFetchLocations() -> Bool {
        return DataConnectionUtilities.shouldPerformNetworkOperation("locationFetchNetworkOption");
    }
    
    @objc public static func shouldPushAttachments() -> Bool {
        return DataConnectionUtilities.shouldPerformNetworkOperation("attachmentPushNetworkOption");
    }
    
    @objc public static func shouldFetchAttachments() -> Bool {
        return DataConnectionUtilities.shouldPerformNetworkOperation("attachmentFetchNetworkOption");
    }
    
    @objc public static func shouldFetchAvatars() -> Bool {
        return DataConnectionUtilities.shouldPerformNetworkOperation("attachmentFetchNetworkOption");
    }
  
}
